nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -1777,12 +1777,282 @@ impl ArrayPrototype {
+    ///
+    /// > Note 1
+    /// >
+    /// > callbackfn should be a function that takes four arguments. reduce
+    /// > calls the callback, as a function, once for each element after the
+    /// > first element present in the array, in ascending order.
+    /// >
+    /// > callbackfn is called with four arguments: the previousValue (value
+    /// > from the previous call to callbackfn), the currentValue (value of the
+    /// > current element), the currentIndex, and the object being traversed.
+    /// > The first time that callback is called, the previousValue and
+    /// > currentValue can be one of two values. If an initialValue was
+    /// > supplied in the call to reduce, then previousValue will be
+    /// > initialValue and currentValue will be the first value in the array.
+    /// > If no initialValue was supplied, then previousValue will be the first
+    /// > value in the array and currentValue will be the second. It is a
+    /// > TypeError if the array contains no elements and initialValue is not
+    /// > provided.
+    /// >
+    /// > reduce does not directly mutate the object on which it is called but
+    /// > the object may be mutated by the calls to callbackfn.
+    /// >
+    /// > The range of elements processed by reduce is set before the first
+    /// > call to callbackfn. Elements that are appended to the array after the
+    /// > call to reduce begins will not be visited by callbackfn. If existing
+    /// > elements of the array are changed, their value as passed to callbackfn
+    /// > will be the value at the time reduce visits them; elements that are
+    /// > deleted after the call to reduce begins and before being visited are
+    /// > not visited.
+    ///
+    /// > Note 2
+    /// >
+    /// > This method is intentionally generic; it does not require that
+    /// > its this value be an Array. Therefore it can be transferred to
+    /// > other kinds of objects for use as a method.
+    fn reduce(agent: &mut Agent, this_value: Value, arguments: ArgumentsList) -> JsResult<Value> {
+        let callback_fn = arguments.get(0);
+        let initial_value = arguments.get(1);
+
+        // 1. Let O be ? ToObject(this value).
+        let o = to_object(agent, this_value)?;
+        // 2. Let len be ? LengthOfArrayLike(O).
+        // NOTE: Guaranteed to be a valid u32 because it's the length of an array-like object.
+        let len = length_of_array_like(agent, o)? as u32;
+
+        // 3. If IsCallable(callbackfn) is false, throw a TypeError exception.
+        let Some(callback_fn) = is_callable(callback_fn) else {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "Callback function is not a function",
+            ));
+        };
+
+        // 4. If len = 0 and initialValue is not present, throw a TypeError exception.
+        if len == 0 && initial_value.is_undefined() {
```

Comment:
**issue:** Spec is unclear with this stuff but "initialValue is not present" does not mean the same as it being `undefined`. You basically need to check the number of arguments and see that there were at least two parameters.

ie. This is okay: `[].reduce(callback, undefined)` but this is not: `[].reduce(callback)`.